src/common/src/hash/consistent_hash/vnode.rs

```diff
@@ -100,3 +106,72 @@ impl VirtualNode {
         (0..Self::COUNT).map(Self::from_index)
     }
 }
+
+impl VirtualNode {
+    pub fn compute_chunk(data_chunk: &DataChunk, keys: &[usize]) -> Vec<VirtualNode> {
+        if let Ok(idx) = keys.iter().exactly_one() &&
+            let ArrayImpl::Serial(serial_array) = data_chunk.column_at(*idx).array_ref() {
+
+            return serial_array.iter()
+                .map(|serial| Self::from_index(extract_vnode_id_from_row_id(serial.unwrap().as_row_id()) as usize))
+                .collect();
+        }
+
+        data_chunk
+            .get_hash_values(keys, Crc32FastBuilder)
+            .into_iter()
+            .map(|hash| hash.to_vnode())
+            .collect_vec()
+    }
+
+    pub fn compute_row(row: &impl Row, indices: &[usize]) -> VirtualNode {
```

Comment:
`&impl Row` also implements `Row` so we can directly...

```suggestion
	/// ....
    pub fn compute_row(row: impl Row, indices: &[usize]) -> VirtualNode {
```